internal/overlord/logstate/manager.go

```diff
@@ -0,0 +1,217 @@
+		buffers:    map[string]*servicelog.RingBuffer{},
+		forwarders: map[serviceTarget]logForwarder{},
+		toAdd:      serviceTargetSet{},
+	}
+}
+
+func (m *logManager) PlanChanged(p *plan.Plan) {
+	m.selectTargets(p)
+	add, remove := deltas(m.planTargets, m.forwarders)
+
+	// Remove obsolete forwarders
+	for st := range remove {
+		m.forwarders[st].Stop()
+		delete(m.forwarders, st)
+	}
+
+	// Add new forwarders
+	m.toAdd.AddAll(add)
+	m.addNewForwarders()
+}
+
+func (m *logManager) selectTargets(p *plan.Plan) {
+	m.planTargets = serviceTargetSet{}
+
+	for serviceName, service := range p.Services {
+		for _, targetName := range service.LogTargets {
+			target, ok := p.LogTargets[targetName]
+			if !ok {
+				logger.Noticef("Log manager cannot find target %q", targetName)
+				continue
+			}
+			if target.Selection != plan.DisabledSelection {
+				m.planTargets.Add(serviceTarget{
+					serviceName: serviceName,
+					targetName:  targetName,
+					target:      *target,
+				})
+			}
+		}
+
+		// Add opt-out targets if none specified
+		if len(service.LogTargets) == 0 {
+			for targetName, target := range p.LogTargets {
+				if target.Selection == plan.OptOutSelection {
+					m.planTargets.Add(serviceTarget{
+						serviceName: serviceName,
+						targetName:  targetName,
+						target:      *target,
+					})
+				}
+			}
+		}
+	}
+}
+
+// Register the service (and buffer) with the log manager, and create new
+// forwarders that are on the manager's to-do list.
+func (m *logManager) NewService(serviceName string, buffer *servicelog.RingBuffer) {
+	m.buffers[serviceName] = buffer
+	// Update forwarders
+	m.addNewForwarders()
+}
+
+// Notifies the log manager that the named service has been stopped, so we can
+// stop any forwarders which are forwarding logs for this service.
+func (m *logManager) StopService(serviceName string) {
+	logger.Debugf("log manager notified of stopping service %q", serviceName)
+	delete(m.buffers, serviceName)
+	for st, p := range m.forwarders {
+		if st.serviceName == serviceName {
+			p.Stop()
+			delete(m.forwarders, st)
+		}
+	}
+}
+
+// addNewForwarders goes through the manager's to-do list (m.toAdd), and checks
+// if each service has been started. If so, create the corresponding forwarder.
+// TODO: should this run in a separate goroutine? so we don't slow down service startups?
+func (m *logManager) addNewForwarders() {
+	stillToAdd := serviceTargetSet{}
+	for st := range m.toAdd {
+		_, started := m.buffers[st.serviceName]
+		if started {
+			// Create new forwarder
+			it := m.buffers[st.serviceName].HeadIterator(0)
+			f, err := newLogForwarder(st.serviceName, st.targetName, st.target, it)
+			if err != nil {
+				logger.Noticef("Cannot forward logs: %v", err)
+			} else {
+				m.forwarders[st] = f
+				logger.Debugf("start pumping %s -> %s/%s",
+					st.serviceName, st.target.Type, st.target.Location)
+				go f.Forward()
+			}
+
+		} else {
+			// Service still not started - to do later
+			stillToAdd.Add(st)
```

Comment:
not sure if this is the best way to manage forwarders.

Reply:
Ok, what's a better way? This is the awkward time **after* the plan has specified that service X should have logs forwarded to target Y, but **before** the daemon has actually started service X. We can't start log forwarding since the service ringbuffer hasn't been created yet. So we make a note to start forwarding later, once we actually receive a reference to the ringbuffer.